WordPress/Classes/ViewRelated/Jetpack/JetpackConnectionWebViewController.swift

```diff
@@ -0,0 +1,311 @@
+            .value?
+            .removingPercentEncoding
+            .flatMap(URL.init(string:))
+    }
+
+    func handleMobileRedirect() {
+        let context = ContextManager.sharedInstance().mainContext
+        let service = BlogService(managedObjectContext: context)
+        let success: () -> Void = { [weak self] in
+            self?.delegate?.jetpackConnectionCompleted()
+        }
+        let failure: (Error) -> Void = { (error) in
+            DDLogError("\(error)")
+            success()
+        }
+        service.syncBlog(
+            blog,
+            success: { [account] in
+                guard let account = account else {
+                    // If there's no account let's pretend this worked
+                    // We don't know what to do, but at least it will dismiss
+                    // the connection flow and refresh the site state
+                    success()
+                    return
+                }
+                service.associateSyncedBlogs(
+                    toJetpackAccount: account,
+                    success: success,
+                    failure: failure
+                )
+            },
+            failure: failure
+        )
+    }
+
+    func performSiteLogin(redirect: URL, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
+        guard let authenticator = WebViewAuthenticator(blog: blog) else {
+                decisionHandler(.allow)
+                return
+        }
+        decisionHandler(.cancel)
+        let request = authenticator.authenticatedRequest(url: redirect)
+        DDLogDebug("Performing site login to \(String(describing: request.url))")
+        pendingSiteRedirect = redirect
+        webView.load(request)
+    }
+
+    func performDotComLogin(redirect: URL) {
+        let context = ContextManager.sharedInstance().mainContext
+        let service = AccountService(managedObjectContext: context)
+        if let account = service.defaultWordPressComAccount(),
+            let token = account.authToken,
+            let username = account.username {
+            authenticateWithDotCom(username: username, token: token, redirect: redirect)
+        } else {
+            presentDotComLogin(redirect: redirect)
+        }
+    }
+
+    func authenticateWithDotCom(username: String, token: String, redirect: URL) {
+        let authenticator = WebViewAuthenticator(credentials: .dotCom(username: username, authToken: token))
+        authenticator.safeRedirect = true
+        let request = authenticator.authenticatedRequest(url: redirect)
+        DDLogDebug("Performing WordPress.com login to \(String(describing: request.url))")
+        webView.load(request)
+    }
+
+    func presentDotComLogin(redirect: URL) {
+        pendingDotComRedirect = redirect
+        observeLoginNotifications(true)
+        SigninHelpers.showLoginForJustWPComFromPresenter(self, forJetpackBlog: blog)
+    }
+
+    func observeLoginNotifications(_ observe: Bool) {
```

Comment:
I'm not crazy about this, I would rather have two methods `startObservingLoginNotifications` and `stopObservingLoginNotifications`. I can see this is similar to what we have in `JetpackLoginViewController` and I like it less there with the return in between, but I'm probably just being nipticky, feel free go completely ignore me on this one.